libcontainer/utils/utils_unix.go

```diff
@@ -98,3 +105,103 @@ func NewSockPair(name string) (parent, child *os.File, err error) {
 	}
 	return os.NewFile(uintptr(fds[1]), name+"-p"), os.NewFile(uintptr(fds[0]), name+"-c"), nil
 }
+
+// WithProcfd runs the passed closure with a procfd path (/proc/self/fd/...)
+// corresponding to the unsafePath resolved within the root. Before passing the
+// fd, this path is verified to have been inside the root -- so operating on it
+// through the passed fdpath should be safe. Do not access this path through
+// the original path strings, and do not attempt to use the pathname outside of
+// the passed closure (the file handle will be freed once the closure returns).
+func WithProcfd(root, unsafePath string, fn func(procfd string) error) error {
+	// Remove the root then forcefully resolve inside the root.
+	unsafePath = stripRoot(root, unsafePath)
+	path, err := securejoin.SecureJoin(root, unsafePath)
+	if err != nil {
+		return fmt.Errorf("resolving path inside rootfs failed: %w", err)
+	}
+
+	procSelfFd, closer := ProcThreadSelf("fd/")
+	defer closer()
+
+	// Open the target path.
+	fh, err := os.OpenFile(path, unix.O_PATH|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return fmt.Errorf("open o_path procfd: %w", err)
+	}
+	defer fh.Close()
+
+	procfd := filepath.Join(procSelfFd, strconv.Itoa(int(fh.Fd())))
+	// Double-check the path is the one we expected.
+	if realpath, err := os.Readlink(procfd); err != nil {
+		return fmt.Errorf("procfd verification failed: %w", err)
+	} else if realpath != path {
+		return fmt.Errorf("possibly malicious path detected -- refusing to operate on %s", realpath)
+	}
+
+	err = fn(procfd)
+	// Make sure the fd stays alive until after the closure returns.
+	runtime.KeepAlive(fh)
```

Comment:
Hmm, this doesn't seem nice. Why was this added? Was it being closed before `fh.Fd()` was used?

I've hit similar issues, but the solutions that I come up with doesn't seem to be possible here.

Reply:
I didn't see it in practice, but the last user of `fh` is a while before we use `procfd` and thus AFAIU the Go GC is entitled to close the file if it happened to run between the two points. The "correct" way of doing this is to use `(*os.File).Sysconn` which guarantees the fd won't be freed but that interface is much uglier.

Reply:
Oh actually, we do `defer fh.Close()` so it should be fine.